man/man3/ConfigFloppy.3
.TH "ConfigFloppy" 3 "Thu Aug 23 2012" "Version 1.0" "FS-UAE Gui for Linux OS" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ConfigFloppy \- 
.PP
This class is responsible for the Floppy Configuration management\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <configfloppy\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBConfigFloppy\fP ()"
.br
.ti -1c
.RI "string \fBgetFloppyDriveConfigStringAt\fP (int i)"
.br
.ti -1c
.RI "string \fBgetFloppyDriveStringAt\fP (int i)"
.br
.ti -1c
.RI "string \fBgetFloppyDriveVolumeConfigString\fP ()"
.br
.RI "\fIRepresents the string of floppy_drive_volume to be written in Config File\&. \fP"
.ti -1c
.RI "string \fBgetFloppyDriveVolumeString\fP ()"
.br
.RI "\fIRepresent the effective value of floppy_drive_volume\&. \fP"
.ti -1c
.RI "string \fBgetFloppyDriveSpeedConfigString\fP ()"
.br
.RI "\fIRepresents the string of floppy_drive_speed to be written in Config File\&. \fP"
.ti -1c
.RI "string \fBgetFloppyDriveSpeedString\fP ()"
.br
.ti -1c
.RI "int \fBgetFloppyImageSize\fP ()"
.br
.RI "\fIReturns the size of the floppy\&. \fP"
.ti -1c
.RI "string \fBgetFloppyImageAt\fP (int position)"
.br
.RI "\fIReturns the string that represet the Floppy Image\&. \fP"
.ti -1c
.RI "void \fBpushBackFloppyImage\fP (string s)"
.br
.RI "\fIAdd new Floppy Image\&. \fP"
.ti -1c
.RI "void \fBeraseFloppyImageAt\fP (int position)"
.br
.RI "\fIErase a Flopy Image\&. \fP"
.ti -1c
.RI "int \fBsetParameter\fP (string parameter, string value)"
.br
.ti -1c
.RI "void \fBsetToDefaultConfiguration\fP ()"
.br
.ti -1c
.RI "bool \fBhasParameter\fP (string parameterName)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "string \fBfloppy_drive\fP [4]"
.br
.ti -1c
.RI "vector< string > \fBfloppy_image\fP"
.br
.RI "\fIIn order to enable floppy­swapping, you must list all floppy images which is to appear in the floppy list\&.
.br
 You can use floppy_image_0, floppy_image_1, floppy_image_2, and so on\&. \fP"
.ti -1c
.RI "string \fBfloppy_drive_volume\fP"
.br
.RI "\fISet volume to 0 to disable floppy drive sounds\&. Max volume is 100\&. The default value is 20\&. \fP"
.ti -1c
.RI "string \fBfloppy_drive_speed\fP"
.br
.RI "\fISet the speed of the emulated floppy drives, in percent\&. The default is 100\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
This class is responsible for the Floppy Configuration management\&. 

In this class are defined all the variables for the correct management of floppy configuration's tab\&.
.br
 In this tab, you can load up to a maximum of 4 physical floppy disk, and create a list with up to 20 floppy images to change the floppy despite having already launched emulation and without having to manually change the configuration file\&. It is also possible to add or remove from the list one or more files at a time\&. At the bottom of the tab there are also two sliders:
.br
 one manages the output volume of the floppy disk drive and is flanked by a checkbox where you can disable the sound (mute);
.br
 the second manages the speed of rotation of the disk\&.
.PP
This class is responsible for the Floppy Configuration\&.
.br
 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "\fBConfigFloppy::ConfigFloppy\fP ()"For example, you can specify 800 to get an 8x increase in speed\&. Use 0 to specify turbo mode\&. 
.SH "Member Function Documentation"
.PP 
.SS "void \fBConfigFloppy::eraseFloppyImageAt\fP (intposition)"
.PP
Erase a Flopy Image\&. 
.SS "string \fBConfigFloppy::getFloppyDriveConfigStringAt\fP (inti)"
.SS "string \fBConfigFloppy::getFloppyDriveSpeedConfigString\fP ()"
.PP
Represents the string of floppy_drive_speed to be written in Config File\&. If floppy_drive_speed is the default value [100], it returns an empty string\&.
.br
 floppy_drive_speed = 100 
.SS "string \fBConfigFloppy::getFloppyDriveSpeedString\fP ()"
.SS "string \fBConfigFloppy::getFloppyDriveStringAt\fP (inti)"
.SS "string \fBConfigFloppy::getFloppyDriveVolumeConfigString\fP ()"
.PP
Represents the string of floppy_drive_volume to be written in Config File\&. If floppy_drive_volume is the default value, it returns an empty string\&.
.br
 floppy_drive_volume = 20 
.SS "string \fBConfigFloppy::getFloppyDriveVolumeString\fP ()"
.PP
Represent the effective value of floppy_drive_volume\&. 
.SS "string \fBConfigFloppy::getFloppyImageAt\fP (intposition)"
.PP
Returns the string that represet the Floppy Image\&. 
.SS "int \fBConfigFloppy::getFloppyImageSize\fP ()"
.PP
Returns the size of the floppy\&. 
.SS "bool \fBConfigFloppy::hasParameter\fP (stringparameterName)"
.SS "void \fBConfigFloppy::pushBackFloppyImage\fP (strings)"
.PP
Add new Floppy Image\&. 
.SS "int \fBConfigFloppy::setParameter\fP (stringparameter, stringvalue)"
.SS "void \fBConfigFloppy::setToDefaultConfiguration\fP ()"
.SH "Member Data Documentation"
.PP 
.SS "string \fBConfigFloppy::floppy_drive\fP[4]\fC [private]\fP"
.SS "string \fBConfigFloppy::floppy_drive_speed\fP\fC [private]\fP"
.PP
Set the speed of the emulated floppy drives, in percent\&. The default is 100\&. 
.SS "string \fBConfigFloppy::floppy_drive_volume\fP\fC [private]\fP"
.PP
Set volume to 0 to disable floppy drive sounds\&. Max volume is 100\&. The default value is 20\&. 
.SS "vector<string> \fBConfigFloppy::floppy_image\fP\fC [private]\fP"
.PP
In order to enable floppy­swapping, you must list all floppy images which is to appear in the floppy list\&.
.br
 You can use floppy_image_0, floppy_image_1, floppy_image_2, and so on\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for FS-UAE Gui for Linux OS from the source code\&.


configfloppy.cpp
/*! \file configfloppy.cpp
 configfloppy.cpp

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2, or (at your option)
 any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 Written by Riva Alessandro Mario, alessandromario.riva@gmail.com
 and Sanvito Davide, dsanvito90@gmail.com. */
#include "configfloppy.h"
#include <sstream>
#include <string>

const string DEFAULTFLOPPYDRIVE = "";
const string DEFAULTFLOPPYDRIVEVOLUME = "20";
const string DEFAULTFLOPPYDRIVESPEED = "100";

ConfigFloppy::ConfigFloppy()
{
    setToDefaultConfiguration();
}

string ConfigFloppy::getFloppyDriveConfigStringAt(int i)
{
    string iString=static_cast<ostringstream*>( &(ostringstream() << i) )->str();
    if (i<0 || i>3){return "";}
    if (floppy_drive[i].compare(DEFAULTFLOPPYDRIVE)==0){return "";}
    else {return "floppy_drive_" + iString + " = " + floppy_drive[i];}
}

string ConfigFloppy::getFloppyDriveStringAt(int i)
{
    if (i<0 || i>3){return "";}
    return floppy_drive[i];
}

string ConfigFloppy::getFloppyDriveVolumeConfigString()
{
    if (floppy_drive_volume.compare(DEFAULTFLOPPYDRIVEVOLUME)==0){return "";}
    else {return "floppy_drive_volume = " + floppy_drive_volume;}
}

string ConfigFloppy::getFloppyDriveVolumeString()
{
    return floppy_drive_volume;
}

string ConfigFloppy::getFloppyDriveSpeedConfigString()
{
    if (floppy_drive_speed.compare(DEFAULTFLOPPYDRIVESPEED)==0){return "";}
    else {return "floppy_drive_speed = " + floppy_drive_speed;}
}

string ConfigFloppy::getFloppyDriveSpeedString()
{
    return floppy_drive_speed;
}

int ConfigFloppy::getFloppyImageSize()
{
    return floppy_image.size();
}

string ConfigFloppy::getFloppyImageAt(int position)
{
    if (position>=(int)floppy_image.size()){return "";}
    return floppy_image.at(position);
}

void ConfigFloppy::pushBackFloppyImage(string s)
{
    floppy_image.push_back(s);
}

void ConfigFloppy::eraseFloppyImageAt(int position)
{
    if (position>=(int)floppy_image.size()){return;}
    floppy_image.erase(floppy_image.begin()+ position);
}
////////////////////////////////////////////////////////////////////////////////////////////////////////////
/** \brief Checks if s is a number
      * \param s is the string to be checked
      * \return true if s is a number
      */
bool static isNumber(string s){
    istringstream buffer(s);
    int intValue;
    buffer >> intValue;
    //la conversione ritorna 0 se non è un numero. Devo stare attento al caso in cui la stringa sia proprio "0"
    if (intValue==0 && s.compare("0")==0) return true;
    return intValue!=0;
}
/** \brief Convert a string to an integer
      * \param s is the string to be converted
      * \return s as integer
      */
int static strToInt(string s){
    istringstream buffer(s);
    int intValue;
    buffer >> intValue;
    return intValue;
}
///////////////////////////////////////////////////////////////////////////////////////////////////////////
int ConfigFloppy::setParameter(string parameter, string value)
{
    if(parameter.compare("floppy_drive_0")==0){
        floppy_drive[0]=value;
    } else if(parameter.compare("floppy_drive_1")==0){
        floppy_drive[1]=value;
    } else if(parameter.compare("floppy_drive_2")==0){
        floppy_drive[2]=value;
    } else if(parameter.compare("floppy_drive_3")==0){
        floppy_drive[3]=value;
    } else if(parameter.compare("floppy_drive_volume")==0){
        if (isNumber(value) && strToInt(value)>=0 && strToInt(value)<=100){
            floppy_drive_volume=value;}
        else{
            floppy_drive_volume=DEFAULTFLOPPYDRIVEVOLUME;
            return -1;
        }
    } else if(parameter.compare("floppy_drive_speed")==0){

        if ((value.compare("0")==0)||(value.compare("100")==0)||(value.compare("200")==0)||(value.compare("300")==0)||(value.compare("400")==0)||(value.compare("500")==0)||(value.compare("600")==0)||(value.compare("700")==0)||(value.compare("800")==0)||(value.compare("900")==0)||(value.compare("1000")==0)||(value.compare("1100")==0)||(value.compare("1200")==0)){
            floppy_drive_speed=value;
        }
        else{
            floppy_drive_speed=DEFAULTFLOPPYDRIVESPEED;
            return -1;
        }

    } else if((parameter.substr(0,parameter.length()-1).compare("floppy_image_")==0 || parameter.substr(0,parameter.length()-2).compare("floppy_image_")==0)&& floppy_image.size()<20){
        floppy_image.push_back(value);
    }
    return 0;
}

void ConfigFloppy::setToDefaultConfiguration()
{
    for(int i=0;i<4;i++){
        floppy_drive[i]=DEFAULTFLOPPYDRIVE;
    }
    floppy_drive_volume=DEFAULTFLOPPYDRIVEVOLUME;
    floppy_drive_speed=DEFAULTFLOPPYDRIVESPEED;
    floppy_image.clear();
}

bool ConfigFloppy::hasParameter(string parameterName)
{
    return (parameterName.substr(0,parameterName.length()-1).compare("floppy_drive_")==0)||(parameterName.substr(0,parameterName.length()-1).compare("floppy_image_")==0)||(parameterName.substr(0,parameterName.length()-2).compare("floppy_image_")==0)||(parameterName.compare("floppy_drive_volume")==0)||(parameterName.compare("floppy_drive_speed")==0);
}
